src/main/java/io/netty/incubator/codec/http3/Http3RequestStreamInboundHandler.java

```diff
@@ -48,33 +47,32 @@ public final boolean isSharable() {
     @Override
     public final void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
         firstFrameReceived = true;
-        if (heldBackMessage != null) {
-            handleMessage(ctx, false);
-        }
-        heldBackMessage = msg;
+        handleBufferedMessage(ctx, false);
+        bufferedMessage = msg;
     }
 
     @Override
     public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
-        if (heldBackMessage != null) {
-            boolean noMoreInput = ((QuicStreamChannel) ctx.channel()).isInputShutdown();
-            handleMessage(ctx, noMoreInput);
-            heldBackMessage = null;
-        }
+        // Once we receive a channelReadComplete we know that we handled all the data that was contained in a
+        // stream frame. At this point we should check if the input was closed and so if this received frame
+        // will be the last on the stream.
+        handleBufferedMessage(ctx, ((QuicStreamChannel) ctx.channel()).isInputShutdown());
+
         super.channelReadComplete(ctx);
     }
 
     @Override
     public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
         // in case the handler is removed between channelRead and channelReadComplete. Should be rare
-        if (heldBackMessage != null) {
-            ReferenceCountUtil.release(heldBackMessage);
-            heldBackMessage = null;
-        }
+        handleBufferedMessage(ctx, ((QuicStreamChannel) ctx.channel()).isInputShutdown());
```

Comment:
imo it should discard. if you remove the handler you probably don't expect more messages from it. in addition, it's not clear whether noMoreInput should be true or false at this point.

Reply:
eg this is correct if the handler is removed just before channelReadComplete, but if it's removed before a further message is read, we will see the same bug where a message is marked as last even though it's not.

Reply:
hmm... true enough... let me fix it